src/ir/possible-contents.cpp

```diff
@@ -1321,12 +1323,504 @@ struct InfoCollector
+        if (auto super = type.getSuperType()) {
+          type = *super;
+        } else {
+          break;
+        }
+      }
+    }
+  }
+
+  doAnalysis([&](Function* func, TNHInfo& info) {
+    // We will need some CFG information below. Computing this is expensive, so
+    // only do it if we find optimization opportunities.
+    std::optional<analysis::CFGBlockIndexes> blockIndexes;
+
+    auto ensureCFG = [&]() {
+      if (!blockIndexes) {
+        auto cfg = analysis::CFG::fromFunction(func);
+        blockIndexes = analysis::CFGBlockIndexes(cfg);
+      }
+    };
+
+    for (auto* call : info.calls) {
+      auto& targetInfo = map[wasm.getFunction(call->target)];
+
+      auto& targetCastParams = targetInfo.castParams;
+      if (targetCastParams.empty()) {
+        continue;
+      }
+
+      // This looks promising, create the CFG if we haven't already, and
+      // optimize.
+      ensureCFG();
+      optimizeCallCasts(
+        call, call->operands, targetCastParams, *blockIndexes, info);
+
+      // Note that we don't need to do anything for targetInfo.traps for a
+      // direct call: the inliner will inline the singleton unreachable in the
+      // target function anyhow.
+    }
+
+    for (auto* call : info.callRefs) {
+      auto targetType = call->target->type;
+      if (!targetType.isRef()) {
+        // This is unreachable or null, and other passes will optimize that.
+        continue;
+      }
+
+      auto iter = typeFunctions.find(targetType.getHeapType());
+      if (iter == typeFunctions.end()) {
+        // No function exists of this type, so the call_ref will trap. We can
+        // mark the target as empty, which has the identical effect.
+        info.inferences[call->target] = PossibleContents::none();
+        continue;
+      }
+
+      // We should only get here in a closed world, in which we know which
+      // functions might be called.
+      assert(options.closedWorld);
+
+      // Go through the targets and ignore any that will trap. That will leave
+      // us with the actually possible targets.
+      const auto& targets = iter->second;
+      std::vector<Function*> possibleTargets;
+      for (Function* target : targets) {
+        auto& targetInfo = map[target];
+        if (targetInfo.traps) {
+          continue;
+        }
+
+        // If our operands will fail a cast, then we will trap.
+        bool traps = false;
+        for (auto& [castIndex, castType] : targetInfo.castParams) {
+          auto operandType = call->operands[castIndex]->type;
+          if (!Type::isSubType(operandType, castType) &&
+              !Type::isSubType(castType, operandType)) {
+            traps = true;
+            break;
+          }
+        }
+        if (!traps) {
+          possibleTargets.push_back(target);
+        }
+      }
+
+      if (possibleTargets.empty()) {
+        // No target is possible.
+        info.inferences[call->target] = PossibleContents::none();
```

Comment:
This inference will constrain the forward analysis. That is, if we say "hey, we've proven nothing can be here" then the forward pass will always avoid putting anything there (because it always takes the intersection with what we say).